### ISSUE TITLE
preg_replace could not parse string

### ISSUE THREAD

**Issue by ging-dev** [phpactor/phpactor#1601]
```
Message: Exception [RuntimeException] preg_replace could not parse string (size 422): Malformed UTF-8 characters, possibly incorrectly encoded at lib/Completion/Core/Util/OffsetHelper.php#15
```
Reproduce: type some Vietnamese characters: à, á, ă...

**Comment by dantleech** [comment]
hi, can you provide some more context? where exactly does this occur? (works for me just entering chars in the editor)

**Comment by ging-dev** [comment]
> hi, can you provide some more context? where exactly does this occur? (works for me just entering chars in the editor)


https://user-images.githubusercontent.com/42226341/168568778-c4ef92c2-e08d-4f8a-8ee4-3f273759c908.mp4

```
[Error - 5:00:07 pm] Request textDocument/completion failed.
  Message: Exception [RuntimeException] preg_replace could not parse string (size 675): Malformed UTF-8 characters, possibly incorrectly encoded at /home/gingdev/code/gingdev/phpactor/lib/Completion/Core/Util/OffsetHelper.php#15
  Code: -32603 
#0 /home/gingdev/code/gingdev/phpactor/lib/Completion/Bridge/TolerantParser/WorseReflection/DoctrineAnnotationCompletor.php(91): Phpactor\Completion\Core\Util\OffsetHelper::lastNonWhitespaceCharacterOffset()
#1 /home/gingdev/code/gingdev/phpactor/lib/Completion/Bridge/TolerantParser/WorseReflection/DoctrineAnnotationCompletor.php(40): Phpactor\Completion\Bridge\TolerantParser\WorseReflection\DoctrineAnnotationCompletor->truncateSource()
#2 /home/gingdev/code/gingdev/phpactor/lib/Completion/Core/ChainCompletor.php(31): Phpactor\Completion\Bridge\TolerantParser\WorseReflection\DoctrineAnnotationCompletor->complete()
#3 /home/gingdev/code/gingdev/phpactor/lib/Completion/Core/Completor/DedupeCompletor.php(27): Phpactor\Completion\Core\ChainCompletor->complete()
#4 /home/gingdev/code/gingdev/phpactor/lib/Extension/LanguageServerCompletion/Handler/CompletionHandler.php(84): Phpactor\Completion\Core\Completor\DedupeCompletor->complete()
#5 [internal function]: Phpactor\Extension\LanguageServerCompletion\Handler\CompletionHandler->Phpactor\Extension\LanguageServerCompletion\Handler\{closure}()
#6 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#7 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#8 /home/gingdev/code/gingdev/phpactor/lib/Extension/LanguageServerCompletion/Handler/CompletionHandler.php(122): Amp\call()
#9 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Handler/HandlerMethodRunner.php(86): Phpactor\Extension\LanguageServerCompletion\Handler\CompletionHandler->completion()
#10 [internal function]: Phpactor\LanguageServer\Core\Handler\HandlerMethodRunner->Phpactor\LanguageServer\Core\Handler\{closure}()
#11 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#12 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#13 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Handler/HandlerMethodRunner.php(102): Amp\call()
#14 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/HandlerMiddleware.php(38): Phpactor\LanguageServer\Core\Handler\HandlerMethodRunner->dispatch()
#15 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\HandlerMiddleware->process()
#16 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/ResponseHandlingMiddleware.php(35): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#17 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\ResponseHandlingMiddleware->process()
#18 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/MethodAliasMiddleware.php(37): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#19 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\MethodAliasMiddleware->process()
#20 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/CancellationMiddleware.php(49): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#21 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\CancellationMiddleware->process()
#22 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/ShutdownMiddleware.php(68): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#23 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\ShutdownMiddleware->process()
#24 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/InitializeMiddleware.php(68): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#25 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\InitializeMiddleware->process()
#26 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/ErrorHandlingMiddleware.php(39): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#27 [internal function]: Phpactor\LanguageServer\Middleware\ErrorHandlingMiddleware->Phpactor\LanguageServer\Middleware\{closure}()
#28 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#29 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#30 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Middleware/ErrorHandlingMiddleware.php(64): Amp\call()
#31 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Middleware/RequestHandler.php(37): Phpactor\LanguageServer\Middleware\ErrorHandlingMiddleware->process()
#32 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Dispatcher/Dispatcher/MiddlewareDispatcher.php(29): Phpactor\LanguageServer\Core\Middleware\RequestHandler->handle()
#33 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(221): Phpactor\LanguageServer\Core\Dispatcher\Dispatcher\MiddlewareDispatcher->dispatch()
#34 [internal function]: Phpactor\LanguageServer\Core\Server\LanguageServer->Phpactor\LanguageServer\Core\Server\{closure}()
#35 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Coroutine.php(67): Generator->current()
#36 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/functions.php(96): Amp\Coroutine->__construct()
#37 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/functions.php(121): Amp\call()
#38 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(238): Amp\asyncCall()
#39 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(212): Phpactor\LanguageServer\Core\Server\LanguageServer->dispatchRequest()
#40 [internal function]: Phpactor\LanguageServer\Core\Server\LanguageServer->Phpactor\LanguageServer\Core\Server\{closure}()
#41 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Coroutine.php(118): Generator->send()
#42 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#43 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Coroutine.php(123): Amp\Coroutine->resolve()
#44 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Internal/Placeholder.php(149): Amp\Coroutine->Amp\{closure}()
#45 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Deferred.php(53): Amp\Promise@anonymous->resolve()
#46 /home/gingdev/code/gingdev/phpactor/vendor/amphp/byte-stream/lib/ResourceInputStream.php(109): Amp\Deferred->resolve()
#47 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Loop/Driver.php(119): Amp\ByteStream\ResourceInputStream::Amp\ByteStream\{closure}()
#48 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Loop/Driver.php(72): Amp\Loop\Driver->tick()
#49 /home/gingdev/code/gingdev/phpactor/vendor/amphp/amp/lib/Loop.php(95): Amp\Loop\Driver->run()
#50 /home/gingdev/code/gingdev/phpactor/vendor/phpactor/language-server/lib/Core/Server/LanguageServer.php(112): Amp\Loop::run()
#51 /home/gingdev/code/gingdev/phpactor/lib/Extension/LanguageServer/Command/StartCommand.php(54): Phpactor\LanguageServer\Core\Server\LanguageServer->run()
#52 /home/gingdev/code/gingdev/phpactor/vendor/symfony/console/Command/Command.php(298): Phpactor\Extension\LanguageServer\Command\StartCommand->execute()
#53 /home/gingdev/code/gingdev/phpactor/vendor/symfony/console/Application.php(1015): Symfony\Component\Console\Command\Command->run()
#54 /home/gingdev/code/gingdev/phpactor/vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application->doRunCommand()
#55 /home/gingdev/code/gingdev/phpactor/lib/Application.php(48): Symfony\Component\Console\Application->doRun()
#56 /home/gingdev/code/gingdev/phpactor/vendor/symfony/console/Application.php(171): Phpactor\Application->doRun()
#57 /home/gingdev/code/gingdev/phpactor/bin/phpactor(47): Symfony\Component\Console\Application->run()
#58 {main}
```

P/S: if character is entered in block `/** */` or `//` then no error will occur.

**Comment by dantleech** [comment]
hm, i can't reproduce this and it's similar to #1391 

i wonder if you're able to reproduce in this test: https://github.com/phpactor/phpactor/blob/master/lib/Completion/Tests/Unit/Core/Util/OffsetHelperTest.php ?

i tried:

```
        yield 'vietnamese chars' => [
            "à, á, ă...<>",
        ];
```

 but it's fine

**Comment by ging-dev** [comment]
Looks like the error is not in vietnamese characters, my text input will go from `af` to `�`  and change to `à`

![image](https://user-images.githubusercontent.com/42226341/168608548-932adf70-2ee8-4a98-990e-508f3315dd32.png)

[debug.log](https://github.com/phpactor/phpactor/files/8700617/debug.log)

**Comment by dantleech** [comment]
hm, i can also open that file (renamed to a PHP file) and edit it without errors

**Comment by ging-dev** [comment]
> hm, i can also open that file (renamed to a PHP file) and edit it without errors

This is actually an invalid input:
![image](https://user-images.githubusercontent.com/42226341/168717093-877928c4-b5c0-41d7-8928-077a5dbfa21d.png)

**Comment by dantleech** [comment]
ok thanks, can reproduce. i guess all we can do is provide a better error?

**Comment by ging-dev** [comment]
> ok thanks, can reproduce. i guess all we can do is provide a better error?

^_^ Do you think `ServerCapabilities.PositionEncoding=utf-8` can solve this problem?

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#serverCapabilities

**Comment by dantleech** [comment]
possibly, think we need to update the LSP protocol to support that (currently at most on 3.16)

**Comment by mamazu** [comment]
Probably also related to #1302 since this also has something to do with mac and utf8

**Comment by dantleech** [comment]
this has been raised again by @axelrinde

I can reproduce this in VSCode, and it seems it can be solved in some cases by returning early from the completors if an UTF-8 error is encountered.

the problem seems to be that the encoding of the file is not UTF-8, VS-Code doesn't covert it before sending (?), or at least it converts it but leaves the invalid chars (?).

Can we confirm that:

- Saving the file fixes the issue

If so, not really sure what the solution here is. Possibly warning when a file is opened if it is not valid UTF-8. We could also "ignore" the errors but at that point the invalid chars are already in Phpactor so it could go wrong in other ways.

**Comment by dantleech** [comment]
Hopefully fixed in #2557
